libass/ass_render.c

```diff
@@ -2834,16 +2885,34 @@ static void add_background(RenderContext *state, EventImages *event_images)
     }
 }
 
+static void setup_shaper(ASS_Shaper *shaper, ASS_Renderer *render_priv)
+{
+    ASS_Track *track = render_priv->track;
+
+    ass_shaper_set_kerning(shaper, track->Kerning);
+    ass_shaper_set_language(shaper, track->Language);
+    ass_shaper_set_level(shaper, render_priv->settings.shaper);
+#ifdef USE_FRIBIDI_EX_API
+    ass_shaper_set_bidi_brackets(shaper,
+            track->parser_priv->feature_flags & FEATURE_MASK(ASS_FEATURE_BIDI_BRACKETS));
+#endif
+    ass_shaper_set_whole_text_layout(shaper,
+            track->parser_priv->feature_flags & FEATURE_MASK(ASS_FEATURE_WHOLE_TEXT_LAYOUT));
+}
+
 /**
  * \brief Main ass rendering function, glues everything together
  * \param event event to render
  * \param event_images struct containing resulting images, will also be initialized
  * Process event, appending resulting ASS_Image's to images_root.
  */
 static bool
-ass_render_event(RenderContext *state, ASS_Event *event,
-                 EventImages *event_images)
+ass_render_event(RenderContext *state, EventImages *event_images)
 {
+    ASS_Event *event = event_images->event;
+
+    event_images->event = NULL;
+
     ASS_Renderer *render_priv = state->renderer;
     if (event->Style >= render_priv->track->n_styles) {
         ass_msg(render_priv->library, MSGL_WARN, "No style found");
```

Comment:
Our log messages are (potentially) passed to a non-thread-safe external function with a pointer to external memory both supplied by API users. Unless we broke API, we need to lock for each `ass_msg` call by default.

Reply:
I don't think holding a lock solves this either, because by the same token, we previously only called the log function _from the thread the user called us from_, and they could hypothetically be relying on some sort of thread-local in there. On the other hand, we never said we'd only call it from one thread, or only once at a time. So, depending on how you think of it (in terms of compatibility with previous undocumented-but-reliable behavior, or compatibility with actual stated guarantees), it's either an API break either way or an API break in neither case.

Reply:
Right, only locking isn't sufficient *(edited previous message to relect this)*.

> On the other hand, we never said we'd only call it from one thread

I don't think a single-threaded lib needs to document all it’s callbacks to be called from a single thread for it to be part of API. Just like even with multi-threading (and potentially an opt-in lock-free `ass_msg`) we probably won't feel the need to point out msg callbacks only happening from the same process context or the same machine.

Reply:
> I don't think marshalling log calls back to the main thread is a reasonable level of complexity to add here. If we really think it's necessary, I think we should just do whatever's necessary to declare this an API break.

I continue to think, we should not silently break API and also be extremely sparing with API breaks.
If there’s a good way¹ to only redirect all logs to the calling process, we should do so. If not and we don't want to properly break API, we could also stick to a single thread by default and communicate `ass_set_threads` perf potential and its accompanying new restriction to our API users.

If we really want to break API though, we’ll likely want to declare an instability period and also tackle [other desirable API adjustments](https://github.com/libass/libass/wiki/ABI-API-changes).